### PR TITLE
Nvcr 6 create edit client page

### PR DESCRIPTION
**Pull request by OstapKH**
Edit client page was implemented
Please, review the code
Looking forward for your feedback and approve

### CODE REVIEW

File: invoicer_project/clients/migrations/0001_initial.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                ('first_name', models.CharField(max_length=35, validators=[django.core.validators.RegexValidator(regex='^[A-Z][a-z]+')])),
+                ('last_name', models.CharField(max_length=35, validators=[django.core.validators.RegexValidator(regex='^[A-Z][a-z]+')])),
+                ('email', models.EmailField(max_length=35, validators=[django.core.validators.RegexValidator(regex='^[a-zA-Z0-9.]{3,20}@(?:[a-zA-Z0-9]{2,20}\\.){1,30}[a-zA-Z]{2,10}$')])),
+                ('phone_number', models.CharField(blank=True, max_length=17, validators=[django.core.validators.RegexValidator(message="Phone number must be entered in the format: '+380679832413'. Up to 15 digits allowed.", regex='^\\+?1?\\d{9,15}$')])),
+                ('zip_code', models.CharField(max_length=10, validators=[django.core.validators.RegexValidator(message="Zip code must be entered in the format: '12345' or '12345-1234'.", regex='^\\d{5}(?:[-\\s]\\d{4})?$')])),
+                ('country', models.CharField(max_length=35, validators=[django.core.validators.RegexValidator(regex='^[A-Z][a-z]+')])),
```
Comment (KateYatsiuk) [inline invoicer_project/clients/migrations/0001_initial.py:27]: @maxchv, sonar has detected the error "[Define a constant instead of duplicating this literal '^[A-Z][a-z]+' 4 times.]" in this file. It is a migration file, the model no longer has this duplication in validation. What should I do about it? Is it possible to change the migration files (and is it a good idea?)? Or maybe we weren't supposed to upload this file to the repository at all?

File: invoicer_project/clients/urls.py [hunk 2]
```diff
@@ -2,14 +2,17 @@
...
 urlpatterns = [
     path("home/", TemplateView.as_view(template_name="clients/clients_list.html"), name='home'),
     path("list/", clients_list, name='clients'),
-    path("client/<int:client_id>", delete_client, name='delete_client'),
-    path("client/", clients_methods, name='clients_add'), 
-    path("add/", TemplateView.as_view(template_name="clients/client_add.html"), name='render_add')
-]
\ No newline at end of file
+    path("client/<int:client_id>", client_crud, name='delete_client'),
+    path("client_edit/<int:client_id>", client_crud, name='edit_client'),
```
Comment (maxchv) [inline invoicer_project/clients/urls.py:13]: do you really need different endpoint? I mean that you have endpoint `/client`.  Why doesn't reuse it?

File: invoicer_project/clients/urls.py [hunk 3]
```diff
@@ -2,14 +2,17 @@
...
+    path("client/", client_crud, name='clients_add'),
+    path("add/", TemplateView.as_view(template_name="clients/client_add.html"), name='render_add'),
+    path("edit/<int:client_id>", TemplateView.as_view(template_name="clients/client_edit.html"), name='render_edit'),
+    path("client_by_id/<int:client_id>", client_crud, name="client_by_id"),
```
Comment (maxchv) [inline invoicer_project/clients/urls.py:17]: The same question

File: invoicer_project/static/js/edit_client_request.js [hunk 4]
```diff
@@ -0,0 +1,328 @@
+const nameSurnMaxLength = 35;
```
Comment (maxchv) [inline invoicer_project/static/js/edit_client_request.js:1]: Sonar argues `Duplicated Lines (%) on New Code 72.6%`. Let's reduce diplications. [Link](https://sonarcloud.io/component_measures?metric=new_duplicated_lines_density&selected=student-project-lnu-2022_invocer%3Ainvoicer_project%2Fstatic%2Fjs%2Fedit_client_request.js&view=list&pullRequest=11&id=student-project-lnu-2022_invocer)

**Review by maxchv — CHANGES_REQUESTED** [review]

File: invoicer_project/clients/models.py [hunk 5]
```diff
@@ -25,13 +26,19 @@ def get_by_zip_code(self, zip_code):
...
+        return self.filter(email=client_email)
+
 
 class Client(models.Model):
     user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='clients')
     first_name = models.CharField(max_length=35, validators=[string_validation])
     last_name = models.CharField(max_length=35, validators=[string_validation])
     email = models.EmailField(max_length=35, validators=[
-        RegexValidator(regex=r'^[a-zA-Z0-9.]{3,20}@(?:[a-zA-Z0-9]{2,20}\.){1,30}[a-zA-Z]{2,10}$')], unique=True)
+        RegexValidator(regex=r'^[a-zA-Z0-9.]{3,20}@(?:[a-zA-Z0-9]{2,20}\.){1,30}[a-zA-Z]{2,10}$')])
```
[the remark this answers is not in the document]
Comment (OstapKH) [reply invoicer_project/clients/models.py:41]: UNIQUE causes problems
For example, if 2 USERS have the same CLIENT, with UNIQUE, it is impossible

File: invoicer_project/invoicer_project/settings.py [hunk 6]
```diff
@@ -119,14 +113,13 @@
...
 
-
 # Static files (CSS, JavaScript, Images)
 # https://docs.djangoproject.com/en/4.1/howto/static-files/
 STATIC_URL = '/static/'
 MEDIA_URL = '/img/'
 
 STATICFILES_DIRS = [
-    BASE_DIR / 'static'
+    os.path.join(BASE_DIR, 'static'),
```
[the remark this answers is not in the document]
Comment (OstapKH) [reply invoicer_project/invoicer_project/settings.py:122]: My fault, I was testing something, and forget to change it.

File: client_edit.html [hunk 7]
```diff
@@ -152,9 +152,6 @@
     </div>
 </div>
 
-
-<!--Jquery connection-->
-<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.1/jquery.min.js"></script>
 <!--M3 elements connection-->
```
[the remark this answers is not in the document]
Comment (OstapKH) [reply client_edit.html]: Done

File: invoicer_project/clients/migrations/0001_initial.py (hunk 1 above)
Comment (KateYatsiuk) [reply invoicer_project/clients/migrations/0001_initial.py:27]: For some reason it generated the same migration file, with duplication. Let's just ignore it if possible

File: invoicer_project/clients/urls.py (hunk 2 above)
Comment (Vitalik2705) [reply invoicer_project/clients/urls.py]: Done

File: invoicer_project/clients/urls.py (hunk 3 above)
Comment (Vitalik2705) [reply invoicer_project/clients/urls.py]: Done

File: invoicer_project/static/js/edit_client_request.js (hunk 4 above)
Comment (Vitalik2705) [reply invoicer_project/static/js/edit_client_request.js]: Done

**Review by NadiiaNanivska — APPROVED** [review]

**Review by Vitalik2705 — APPROVED** [review]

File: invoicer_project/clients/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (maxchv) [reply invoicer_project/clients/models.py:41]: What is different between user and client?

File: invoicer_project/static/js/utils_clients.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
+export const host = "http://127.0.0.1:8000";
```
Comment (maxchv) [inline invoicer_project/static/js/utils_clients.js:1]: I think, that this file should have other name than `utils_clients.js`. It is good idea to move general functions to one place, but let's make next step to improve our code. Let's gather functions related to validations to one place, functions related to interaction with server - to another place and other functions could be left here.

**Review by maxchv — CHANGES_REQUESTED** [review]
It is need to make one additional step to improve our code

File: invoicer_project/static/js/utils_clients.js (hunk 8 above)
Comment (Vitalik2705) [reply invoicer_project/static/js/utils_clients.js:2]: Done

**Review by NadiiaNanivska — APPROVED** [review]

**Review by KateSydor — APPROVED** [review]

**Review by KateYatsiuk — APPROVED** [review]

**Review by dmitru4ok — APPROVED** [review]

**Review by maxchv — APPROVED** [review]
